accepted/2025/PrunePackageReference-with-direct-PackageReference.md

```diff
@@ -0,0 +1,264 @@
+    <PackageReference Condition="'$(TargetFramework)' == 'net9.0'" Include="System.IO.Pipelines" Version="9.0.2" />
+  </ItemGroup>
+```
+
+In this situation, System.IO.Pipelines was added to the framework in .NET 9, so the .NET 10 and .NET 9 references are not necessary.
+Given that the package is unnecessary in *every* framework it's been added in, we'd want the customer to remove both conditional PackageReference items.
+
+### How do we tell the customers we want them to remove a PackageReference
+
+We raise a NU1510 warning.
+
+For multi-targeted projects, the warning will be raised if at least one of the frameworks is .NET 10.
+We will raise an NU1510 when the PackageReference can be completely removed from the project file.
+In other words, when a package would be pruned for every target framework it is specified in, raise NU1510.
+
+In order to provide continuity, and minimize disruption that would be caused by raising NU1510 for existing projects, as per the [Breaking Change guidelines](https://github.com/dotnet/sdk/blob/main/documentation/project-docs/breaking-change-guidelines.md#tie-potentially-impactful-changes-to-the-target-tfm) NU1510 will only be raised for projects targeting net10, requiring an explicit customer action to get this diagnostic.
+
+Referring back to the [scenarios](#scenarios-for-direct-packagereference-items) listed above:
+
+| Scenario | NU1510 | Notes |
+|----------|--------|-------|
+| 1 | Yes | |
+| 2a | No | The warning may be beneficial, but complex heuristic would be harder to reason about. |
+| 2b | No | Eventual runtime failure risks |
+| 2c | No | Eventual runtime failure risks |
+| 2d | Yes | Removal of the PR leads to the package being completely gone from the project. |
+
+### Direct PackageReference in pruning range
+
+The primary motivation for pruning is that the audit functionality and external scanners no longer sees those package references and raise false positives.
+Transitively pruned packages do not appear in the assets file as dependencies of the packages that originally referenced them.
+They have effectively been selected as unnecessary during the framework selection step of dependency resolution.
+The only place the id appears in the assets file is the `project` section under the `packagesToPrune`.
+
+When the NU1510 warning is raised, we are giving the customer a clear signal that they need to remove the PackageReference.
+In these scenarios, what happens to the direct package references is not very critical.
+
+In the partial prune scenarios such as 2b and 2c, pruning directs like transitive could be beneficial in the .NET leg, helping avoid additional audit warnings.
+However, due to the fact that the PackageReference is in other frameworks, the benefit is more minimal, since if the package itself had a vulnerability, the audit warning will be warranted since the customer is using it in a case where it is not being pruned.
```

Comment:
This is a very good point that I don't remember us discussing before.  This convinces me that direct pruning has minimal benefit.

Reply:
> This is a very good point that I don't remember us discussing before

We did not discuss it yeah. It was an idea that I was able to clearly put together after I rewatch the recording of the meeting we had for direct pruning.

> Conflict resolution should still apply in this case. @dsplaisted we may want to test that and ensure conflict resolution will follow the same rules as pruning, or at least document the behavior.

In the current proposal I'm applying `IncludeAssets=None` in https://github.com/NuGet/NuGet.Client/pull/6545. 
The benefit of that is that there's no conflict resolution to be done. The con is the aliasing changes. 

If we want to allow conflict resolution to handle it, we *could* just do PrivateAssets=all, but that'll make things like analyzers apply long term.